src/get-started/cryptocurrency-advanced.md

```diff
@@ -0,0 +1,835 @@
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+pub struct WalletQuery {
+    /// Public key of the queried wallet.
+    pub pub_key: PublicKey,
+}
+```
+
+Besides this we also declare a number of specific structures. Said
+structures will be used for generating responses to users’ requests:
+
+```rust
+/// The structure returned by the REST API.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TransactionResponse {
+    /// Hash of the transaction.
+    pub tx_hash: Hash,
+}
+
+/// Proof of existence for specific wallet.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WalletProof {
+    /// Proof to the whole database table.
+    pub to_table: MapProof<Hash, Hash>,
+    /// Proof to the specific wallet in this table.
+    pub to_wallet: MapProof<PublicKey, Wallet>,
+}
+
+/// Wallet history.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WalletHistory {
+    pub proof: ListProof<Hash>,
+    pub transactions: Vec<WalletTransactions>,
+}
+```
+
+The following structure is based on the two previous auxiliary structures. It is
+this structure, that a user receives as
+a response to his request. The structure contains the necessary information on
+the wallet together with the corresponding
+proofs of existence thereof and the correctness of its history.
+
+```rust
+/// Wallet information.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WalletInfo {
+    pub block_proof: BlockProof,
+    pub wallet_proof: WalletProof,
+    pub wallet_history: Option<WalletHistory>,
+}
+```
+
+Now let’s define the method that will allow us to obtain information on a
+particular wallet together with cryptographic
+proof of its existence. The proofs also allow to get confirmation of existence
+of a particular transaction in the wallet history.
+
+First, we claim `API` structure to which the mentioned method refers. We create
+two schemas – one general schema of our
+blockchain and the other one is the schema of our service described in the
+relevant section above:
+
+```rust
+impl CryptocurrencyApi {
+    pub fn wallet_info(
+      state: &ServiceApiState,
+      query: WalletQuery
+    ) -> api::Result<WalletInfo> {
+        let snapshot = state.snapshot();
+        let general_schema = blockchain::Schema::new(&snapshot);
+        let currency_schema = CurrencySchema::new(&snapshot);
+```
+
+Secondly, we get the current height of our blockchain and obtain all the blocks
+and their precommits to start building the proof:
+
+```rust
+let max_height = general_schema.block_hashes_by_height().len() - 1;
+
+let block_proof = general_schema
+    .block_and_precommits(Height(max_height))
+    .unwrap();
+```
+
+We then obtain the first part of the proof - from the state hash of our service
+to the `Wallets` table:
+
+```rust
+let to_table: MapProof<Hash, Hash> =
+    general_schema.get_proof_to_service_table(CRYPTOCURRENCY_SERVICE_ID, 0);
+```
+
+Note that we indicate `0` as the value of the upper boarder of the interval.
```

Comment:
What is "the upper boarder of the interval"? I'd suggest to call this value "the table identifier" (or "index identifier").